bindings/src/main/java/io/pravega/storage/hdfs/HDFSChunkStorage.java

```diff
@@ -0,0 +1,360 @@
+                return 0;
+            }
+
+            // We need to be very careful with IOUtils.copyBytes. There are many overloads with very similar signatures.
+            // There is a difference between (InputStream, OutputStream, int, boolean) and (InputStream, OutputStream, long, boolean),
+            // in that the one with "int" uses the third arg as a buffer size, and the one with "long" uses it as the number
+            // of bytes to copy.
+            IOUtils.copyBytes(data, stream, (long) length, false);
+
+            stream.flush();
+        } catch (IOException e) {
+            throwException(handle.getChunkName(), "doWrite", e);
+        }
+        return length;
+    }
+
+    @Override
+    public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException, UnsupportedOperationException {
+        ensureInitializedAndNotClosed();
+        int length = 0;
+        try {
+            val sources = new Path[chunks.length - 1];
+            for (int i = 1; i < chunks.length; i++) {
+                val chunkLength = chunks[i].getLength();
+                this.fileSystem.truncate(getFilePath(chunks[i].getName()), chunkLength);
+                sources[i - 1] = getFilePath(chunks[i].getName());
+                length += chunkLength;
+            }
+            // Concat source file into target.
+            this.fileSystem.concat(getFilePath(chunks[0].getName()), sources);
+        } catch (IOException e) {
+            throwException(chunks[0].getName(), "doConcat", e);
+        }
+        return length;
+    }
+
+    @Override
+    protected boolean doTruncate(ChunkHandle handle, long offset) throws ChunkStorageException, UnsupportedOperationException {
+        throw new UnsupportedOperationException(getClass().getName() + " does not support chunk truncation.");
+    }
+
+    @Override
+    protected boolean doSetReadOnly(ChunkHandle handle, boolean isReadOnly) throws ChunkStorageException, UnsupportedOperationException {
+        try {
+            this.fileSystem.setPermission(getFilePath(handle.getChunkName()), isReadOnly ? READONLY_PERMISSION : READWRITE_PERMISSION);
+            return true;
+        } catch (IOException e) {
+            throwException(handle.getChunkName(), "doSetReadOnly", e);
+        }
+        return false;
+    }
+
+    //endregion
+
+    //region Storage Implementation
+
+    @SneakyThrows(IOException.class)
+    public void initialize() {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkState(this.fileSystem == null, "HDFSStorage has already been initialized.");
+        Configuration conf = new Configuration();
+        conf.set("fs.default.name", this.config.getHdfsHostURL());
+        conf.set("fs.default.fs", this.config.getHdfsHostURL());
+        conf.set("fs.hdfs.impl", "org.apache.hadoop.hdfs.DistributedFileSystem");
+
+        // FileSystem has a bad habit of caching Clients/Instances based on target URI. We do not like this, since we
```

Comment:
Fix this comment as the last part does not read correctly.

Reply:
fixed